File: expired/lingua_1-main/lingua_1-main/electron_node/docs/AGGREGATOR/aggregator_decision.ts
/* Aggregator core decision logic: Text Incompleteness Score + Language Stability Gate
   Copy-paste friendly. No external deps.
*/

export type Mode = "offline" | "room";
export type StreamAction = "MERGE" | "NEW_STREAM";

export interface LangProbs {
  top1: string;
  p1: number;
  top2?: string;
  p2?: number;
}

export interface UtteranceInfo {
  text: string;
  startMs: number;
  endMs: number;
  lang: LangProbs;
  qualityScore?: number;
  isFinal?: boolean;
  isManualCut?: boolean;
}

export interface AggregatorTuning {
  strongMergeMs: number;
  softGapMs: number;
  hardGapMs: number;

  langStableP: number;
  langSwitchMargin: number;
  langSwitchRequiresGapMs: number;

  scoreThreshold: number;
  wShort: number;
  wVeryShort: number;
  wGapShort: number;
  wNoStrongPunct: number;
  wEndsWithConnective: number;
  wLowQuality: number;
  lowQualityThreshold: number;

  shortCjkChars: number;
  veryShortCjkChars: number;
  shortEnWords: number;
  veryShortEnWords: number;

  commitIntervalMs: number;
  commitLenCjk: number;
  commitLenEnWords: number;
}

export function defaultTuning(mode: Mode): AggregatorTuning {
  const isRoom = mode === "room";
  return {
    strongMergeMs: isRoom ? 600 : 700,
    softGapMs: isRoom ? 1000 : 1500,
    hardGapMs: isRoom ? 1500 : 2000,

    langStableP: 0.8,
    langSwitchMargin: isRoom ? 0.18 : 0.15,
    langSwitchRequiresGapMs: isRoom ? 500 : 600,

    scoreThreshold: 3,
    wShort: 2,
    wVeryShort: 3,
    wGapShort: 2,
    wNoStrongPunct: 1,
    wEndsWithConnective: 1,
    wLowQuality: 1,
    lowQualityThreshold: isRoom ? 0.5 : 0.45,

    shortCjkChars: isRoom ? 9 : 10,
    veryShortCjkChars: 4,
    shortEnWords: isRoom ? 5 : 6,
    veryShortEnWords: 3,

    commitIntervalMs: isRoom ? 900 : 1400,
    commitLenCjk: isRoom ? 22 : 30,
    commitLenEnWords: isRoom ? 10 : 12,
  };
}

export function decideStreamAction(
  prev: UtteranceInfo | null,
  curr: UtteranceInfo,
  mode: Mode,
  tuning: AggregatorTuning = defaultTuning(mode)
): StreamAction {
  if (!prev) return "NEW_STREAM";

  const gapMs = Math.max(0, curr.startMs - prev.endMs);

  // Hard rules
  if (curr.isFinal || curr.isManualCut) return "NEW_STREAM";
  if (gapMs >= tuning.hardGapMs) return "NEW_STREAM";

  // Language stability gate
  if (isLangSwitchConfident(prev.lang, curr.lang, gapMs, tuning)) return "NEW_STREAM";

  // Strong merge
  if (gapMs <= tuning.strongMergeMs) return "MERGE";

  const score = textIncompletenessScore(prev, curr, gapMs, tuning);
  if (score >= tuning.scoreThreshold && gapMs <= tuning.softGapMs) return "MERGE";

  return "NEW_STREAM";
}

export function isLangSwitchConfident(
  prevLang: LangProbs,
  currLang: LangProbs,
  gapMs: number,
  tuning: AggregatorTuning
): boolean {
  if (gapMs <= tuning.langSwitchRequiresGapMs) return false;
  if (prevLang.p1 < tuning.langStableP || currLang.p1 < tuning.langStableP) return false;
  if (prevLang.top1 === currLang.top1) return false;

  const p2 = currLang.p2 ?? 0;
  return (currLang.p1 - p2) >= tuning.langSwitchMargin;
}

export function textIncompletenessScore(
  prev: UtteranceInfo,
  curr: UtteranceInfo,
  gapMs: number,
  tuning: AggregatorTuning
): number {
  let score = 0;

  const isCjk = looksLikeCjk(curr.text);
  const cjkChars = isCjk ? countCjkChars(curr.text) : 0;
  const enWords = !isCjk ? countWords(curr.text) : 0;

  const short = isCjk ? cjkChars < tuning.shortCjkChars : enWords < tuning.shortEnWords;
  const veryShort = isCjk ? cjkChars < tuning.veryShortCjkChars : enWords < tuning.veryShortEnWords;

  if (veryShort) score += tuning.wVeryShort;
  else if (short) score += tuning.wShort;

  if (gapMs < (tuning.strongMergeMs + 200)) score += tuning.wGapShort;

  if (!endsWithStrongSentencePunct(curr.text)) score += tuning.wNoStrongPunct;

  if (endsWithConnectiveOrFiller(curr.text)) score += tuning.wEndsWithConnective;

  const q = curr.qualityScore ?? 1.0;
  if (q < tuning.lowQualityThreshold) score += tuning.wLowQuality;

  if (!endsWithStrongSentencePunct(prev.text) && gapMs <= tuning.softGapMs) score += 1;

  return score;
}

export function shouldCommit(
  pendingText: string,
  lastCommitTsMs: number,
  nowMs: number,
  mode: Mode,
  tuning: AggregatorTuning = defaultTuning(mode)
): boolean {
  const elapsed = nowMs - lastCommitTsMs;
  if (elapsed >= tuning.commitIntervalMs) return true;

  const isCjk = looksLikeCjk(pendingText);
  if (isCjk) return countCjkChars(pendingText) >= tuning.commitLenCjk;
  return countWords(pendingText) >= tuning.commitLenEnWords;
}

/* Helpers */

export function endsWithStrongSentencePunct(s: string): boolean {
  const t = s.trim();
  if (!t) return false;
  return /[。！？.!?；;]$/.test(t);
}

export function looksLikeCjk(s: string): boolean {
  return /[\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uac00-\ud7af]/.test(s);
}

export function countCjkChars(s: string): number {
  const m = s.match(/[\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uac00-\ud7af]/g);
  return m ? m.length : 0;
}

export function countWords(s: string): number {
  const t = s.trim();
  if (!t) return 0;
  return t.split(/\s+/).filter(Boolean).length;
}

export function endsWithConnectiveOrFiller(s: string): boolean {
  const t = s.trim().toLowerCase();
  if (!t) return false;

  const en = ["and", "but", "so", "because", "then"];
  const zh = ["然后", "所以", "但是", "就是", "那个", "嗯", "呃"];
  const ja = ["で", "から", "けど", "えっと"];
  const ko = ["그리고", "근데", "그래서", "어", "음"];

  for (const w of en) if (t.endsWith(" " + w) || t === w) return true;
  for (const w of zh) if (t.endsWith(w)) return true;
  for (const w of ja) if (t.endsWith(w)) return true;
  for (const w of ko) if (t.endsWith(w)) return true;

  return false;
}

File: expired/lingua_1-main/lingua_1-main/electron_node/docs/AGGREGATOR/aggregator_decision.rs
// Aggregator core decision logic: Text Incompleteness Score + Language Stability Gate
// Copy-paste friendly. No external deps.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Mode {
    Offline,
    Room,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StreamAction {
    Merge,
    NewStream,
}

#[derive(Debug, Clone)]
pub struct LangProbs {
    pub top1: String,
    pub p1: f32,
    pub top2: Option<String>,
    pub p2: Option<f32>,
}

#[derive(Debug, Clone)]
pub struct UtteranceInfo {
    pub text: String,
    pub start_ms: i64,
    pub end_ms: i64,
    pub lang: LangProbs,
    pub quality_score: Option<f32>,
    pub is_final: bool,
    pub is_manual_cut: bool,
}

#[derive(Debug, Clone)]
pub struct AggregatorTuning {
    pub strong_merge_ms: i64,
    pub soft_gap_ms: i64,
    pub hard_gap_ms: i64,

    pub lang_stable_p: f32,
    pub lang_switch_margin: f32,
    pub lang_switch_requires_gap_ms: i64,

    pub score_threshold: i32,
    pub w_short: i32,
    pub w_very_short: i32,
    pub w_gap_short: i32,
    pub w_no_strong_punct: i32,
    pub w_ends_with_connective: i32,
    pub w_low_quality: i32,
    pub low_quality_threshold: f32,

    pub short_cjk_chars: usize,
    pub very_short_cjk_chars: usize,
    pub short_en_words: usize,
    pub very_short_en_words: usize,

    pub commit_interval_ms: i64,
    pub commit_len_cjk: usize,
    pub commit_len_en_words: usize,
}

pub fn default_tuning(mode: Mode) -> AggregatorTuning {
    let is_room = matches!(mode, Mode::Room);
    AggregatorTuning {
        strong_merge_ms: if is_room { 600 } else { 700 },
        soft_gap_ms: if is_room { 1000 } else { 1500 },
        hard_gap_ms: if is_room { 1500 } else { 2000 },

        lang_stable_p: 0.80,
        lang_switch_margin: if is_room { 0.18 } else { 0.15 },
        lang_switch_requires_gap_ms: if is_room { 500 } else { 600 },

        score_threshold: 3,
        w_short: 2,
        w_very_short: 3,
        w_gap_short: 2,
        w_no_strong_punct: 1,
        w_ends_with_connective: 1,
        w_low_quality: 1,
        low_quality_threshold: if is_room { 0.50 } else { 0.45 },

        short_cjk_chars: if is_room { 9 } else { 10 },
        very_short_cjk_chars: 4,
        short_en_words: if is_room { 5 } else { 6 },
        very_short_en_words: 3,

        commit_interval_ms: if is_room { 900 } else { 1400 },
        commit_len_cjk: if is_room { 22 } else { 30 },
        commit_len_en_words: if is_room { 10 } else { 12 },
    }
}

pub fn decide_stream_action(
    prev: Option<&UtteranceInfo>,
    curr: &UtteranceInfo,
    _mode: Mode,
    tuning: &AggregatorTuning,
) -> StreamAction {
    let Some(prev) = prev else { return StreamAction::NewStream; };

    let gap_ms = (curr.start_ms - prev.end_ms).max(0);

    if curr.is_final || curr.is_manual_cut { return StreamAction::NewStream; }
    if gap_ms >= tuning.hard_gap_ms { return StreamAction::NewStream; }

    if is_lang_switch_confident(&prev.lang, &curr.lang, gap_ms, tuning) {
        return StreamAction::NewStream;
    }

    if gap_ms <= tuning.strong_merge_ms { return StreamAction::Merge; }

    let score = text_incompleteness_score(prev, curr, gap_ms, tuning);

    if score >= tuning.score_threshold && gap_ms <= tuning.soft_gap_ms {
        StreamAction::Merge
    } else {
        StreamAction::NewStream
    }
}

pub fn is_lang_switch_confident(
    prev: &LangProbs,
    curr: &LangProbs,
    gap_ms: i64,
    tuning: &AggregatorTuning,
) -> bool {
    if gap_ms <= tuning.lang_switch_requires_gap_ms { return false; }
    if prev.p1 < tuning.lang_stable_p || curr.p1 < tuning.lang_stable_p { return false; }
    if prev.top1 == curr.top1 { return false; }
    let p2 = curr.p2.unwrap_or(0.0);
    (curr.p1 - p2) >= tuning.lang_switch_margin
}

pub fn text_incompleteness_score(
    prev: &UtteranceInfo,
    curr: &UtteranceInfo,
    gap_ms: i64,
    tuning: &AggregatorTuning,
) -> i32 {
    let mut score: i32 = 0;

    let is_cjk = looks_like_cjk(&curr.text);
    let cjk_chars = if is_cjk { count_cjk_chars(&curr.text) } else { 0 };
    let en_words = if !is_cjk { count_words(&curr.text) } else { 0 };

    let short = if is_cjk { cjk_chars < tuning.short_cjk_chars } else { en_words < tuning.short_en_words };
    let very_short = if is_cjk { cjk_chars < tuning.very_short_cjk_chars } else { en_words < tuning.very_short_en_words };

    if very_short { score += tuning.w_very_short; }
    else if short { score += tuning.w_short; }

    if gap_ms < (tuning.strong_merge_ms + 200) { score += tuning.w_gap_short; }

    if !ends_with_strong_sentence_punct(&curr.text) { score += tuning.w_no_strong_punct; }

    if ends_with_connective_or_filler(&curr.text) { score += tuning.w_ends_with_connective; }

    let q = curr.quality_score.unwrap_or(1.0);
    if q < tuning.low_quality_threshold { score += tuning.w_low_quality; }

    if !ends_with_strong_sentence_punct(&prev.text) && gap_ms <= tuning.soft_gap_ms { score += 1; }

    score
}

pub fn should_commit(
    pending_text: &str,
    last_commit_ts_ms: i64,
    now_ms: i64,
    tuning: &AggregatorTuning,
) -> bool {
    let elapsed = now_ms - last_commit_ts_ms;
    if elapsed >= tuning.commit_interval_ms { return true; }

    let is_cjk = looks_like_cjk(pending_text);
    if is_cjk {
        count_cjk_chars(pending_text) >= tuning.commit_len_cjk
    } else {
        count_words(pending_text) >= tuning.commit_len_en_words
    }
}

/* Helpers */

pub fn ends_with_strong_sentence_punct(s: &str) -> bool {
    let t = s.trim_end();
    if t.is_empty() { return false; }
    matches!(t.chars().last().unwrap(), '。' | '！' | '？' | '.' | '!' | '?' | '；' | ';')
}

pub fn looks_like_cjk(s: &str) -> bool {
    s.chars().any(|c| {
        let u = c as u32;
        (0x3040..=0x30FF).contains(&u)
            || (0x3400..=0x4DBF).contains(&u)
            || (0x4E00..=0x9FFF).contains(&u)
            || (0xAC00..=0xD7AF).contains(&u)
    })
}

pub fn count_cjk_chars(s: &str) -> usize {
    s.chars().filter(|c| {
        let u = *c as u32;
        (0x3040..=0x30FF).contains(&u)
            || (0x3400..=0x4DBF).contains(&u)
            || (0x4E00..=0x9FFF).contains(&u)
            || (0xAC00..=0xD7AF).contains(&u)
    }).count()
}

pub fn count_words(s: &str) -> usize {
    s.split_whitespace().filter(|w| !w.is_empty()).count()
}

pub fn ends_with_connective_or_filler(s: &str) -> bool {
    let t = s.trim().to_lowercase();
    if t.is_empty() { return false; }

    let en = ["and", "but", "so", "because", "then"];
    for w in en {
        if t == w || t.ends_with(&format!(" {}", w)) { return true; }
    }

    let zh = ["然后", "所以", "但是", "就是", "那个", "嗯", "呃"];
    let ja = ["で", "から", "けど", "えっと"];
    let ko = ["그리고", "근데", "그래서", "어", "음"];

    for w in zh.iter().chain(ja.iter()).chain(ko.iter()) {
        if t.ends_with(w) { return true; }
    }
    false
}
